aws/s3.go

```diff
@@ -0,0 +1,306 @@
+package aws
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	awsgo "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/gruntwork-io/cloud-nuke/logging"
+	"github.com/gruntwork-io/gruntwork-cli/errors"
+)
+
+// getS3BucketRegion returns S3 Bucket region.
+func getS3BucketRegion(svc *s3.S3, bucketName *string) (string, error) {
+	input := &s3.GetBucketLocationInput{
+		Bucket: bucketName,
+	}
+
+	result, err := svc.GetBucketLocation(input)
+	if err != nil {
+		return "", err
+	}
+
+	if result.LocationConstraint == nil {
+		// GetBucketLocation returns nil for us-east-1
+		// https://github.com/aws/aws-sdk-go/issues/1687
+		return "us-east-1", nil
+	}
+	return *result.LocationConstraint, nil
+}
+
+// getS3BucketTags returns S3 Bucket tags.
+func getS3BucketTags(svc *s3.S3, bucketName *string) ([]map[string]string, error) {
+	input := &s3.GetBucketTaggingInput{
+		Bucket: bucketName,
+	}
+
+	tags := []map[string]string{}
+
+	// Please note that svc argument should be created from a session object which is
+	// in the same region as the bucket or GetBucketTagging will fail.
+	result, err := svc.GetBucketTagging(input)
+	if err != nil {
+		return tags, err
+	}
+
+	for _, tagSet := range result.TagSet {
+		tags = append(tags, map[string]string{"Key": *tagSet.Key, "Value": *tagSet.Value})
+	}
+
+	return tags, nil
+}
+
+// hasValidTags checks if bucket tags permit it to be in the deletion list.
+func hasValidTags(svc *s3.S3, bucket *s3.Bucket) (bool, string) {
+	// Exclude deletion of any buckets with cloud-nuke-excluded tags
+	bucketTags, err := getS3BucketTags(svc, bucket.Name)
+	if len(bucketTags) > 0 {
+		for _, tagSet := range bucketTags {
+			if tagSet["Key"] == "cloud-nuke-excluded" && tagSet["Value"] == "true" {
```

Comment:
More of a general question / idea this gave me: It seems [tags are case sensitive](https://docs.aws.amazon.com/AmazonS3/latest/user-guide/add-object-tags.html) - perhaps we should eventually add (probably somewhere more centralized) a standard lookup feature that normalizes keys so that even if someone accidentally capitalized this tag like `Cloud-nuke-excluded` it would still have the desired effect of protecting the bucket.

Reply:
Makes sense. Will convert it to lower and then compare key and value. Ideally when resources start supporting tags - the tag comparison function should be central. Right now it is local to S3.

Reply:
Done as per https://github.com/gruntwork-io/cloud-nuke/pull/96/commits/9e05a356693d5d9bdb212c57ef8fc19dd3bf9909